FIX: problem when closing settings fig

juangpc opened the pull request:
Hi, 

This PR attempts to fix #78 (opened by @jasmainak).

There are two small problems/typos and a QEvent object which is not sent to the closeEvent call.

agramfort commented:
I confirm it fixes the problem. Any chance we could add a test for this?

thx @juangpc !

juangpc commented:
Thank you. 
Yes, I'll see what I can do to test this menu interaction.

juangpc commented:
Hello again, 

Following @agramfort comment, I'm trying to provide additional test coverage. 

The tests I've added would've shattered the original issue (#78) so that is progress. However there are some comments, questions I'd like to share: 

 - I'm trying to keep all the structure of the code as I see it now. I see that functionality of the mne-qt-browser is now being tested in the ```test_raw.py``` script inside the ```mne-python``` repo. Although some helpers functions are added as members to the ```class PyQtGraphBrowser``` which is in ```_py_figure.py``` inside the ```mne-qt-browser``` repo. 
 Even in this repository documentation, when explaining how to do the tests, one needs to go to run them to the ```mne-python``` repository. So I'm trying to follow this structure. 
- There are thus two linked necessary PRs. One being this one, and a second one on the ```mne-python``` repo. I haven't created it yet. I'll wait to see what is best. It will come from [this](https://github.com/juangpc/mne-python/commit/90399707174808b52e3ec4b23a5ed181bf79fec0) branch on my fork. Should I create it?
- I've added the mark ```pgtest``` so that it doesn't interact with the matplotlib related classes. Is this correct?
- I've added the QTest dependency for the ```test_raw.py``` script in mne-python repo. Is this ok? It works in my env.
- Added additional tests to cover all the members (QActions) on the toolbar of the mne-qt-browser. Should I create a separate PR? 
- None of my branches are rebased. Should I? I hear you guys rebase and later squash. Checked the documentation and contribution guide I couldn't find anything related.

Thanks.

agramfort commented:
@juangpc can you link the PR here I don't see it?

there are some tests run here. Any chance to add a test here?

juangpc commented:
Hi, 

Thanks for your mention @hoechenberger.

 - I (think I) have followed @agramfort guidance to place here all the tests. I've increased the intensity and granularity with which the GUI components are tested. My intention is to eventually lead to a 100% coverage.

 - There is this sensitivity slider in the settings dialog issue I've created (#103). I don't know exactly what it is supposed to do, but it doesn't seem to be doing too much. The effect is similar to widening the shown time span. In any case, the test, does test the slider as a GUI component, and its effect in the model. This component is neighboring the Settings dialog, which is the original subject of this PR. I think it can belong to the same PR. 

 - I've added a ',' comma shortcut for the settings dialog. It was quite useful during development and I figure it might become useful for the user. Many SW applications use this shortcut for settings. (Again) feel free to delete this shortcut. No hurt feelings. 

 - Changed the order of the subsampling methods because 'mean' seems more convenient as a default method, (index 0).

 - Since the tests are now migrating towards this repository, I've edited the README file too. 
 
Thanks,

juangpc commented:
Ready for merge on my side.

hoechenberger commented:
Hello @juangpc, thanks for the effort you're putting in here!!

I'm just a little concerned this work is getting a bit outside of the scope of the original idea of this PR, and I wonder if it would make sense to split it into smaller, separate PRs that each basically do one thing only. This will make the review process faster and easier, and allows for a better discussion of the individual changes you're proposing!

For example, while I do like the idea of adding a shortcut for settings, it should conform to the system standards, e.g. Cmd+, on macOS, and to whatever the defaults might be in windows and Linux.

juangpc commented:
Ok. No worries. I’ll drop that. Thanks anyway.

On Thu, Apr 7, 2022 at 1:36 AM Richard Höchenberger <
***@***.***> wrote:

> Hello @juangpc <https://github.com/juangpc>, thanks for the effort you're
> putting in here!!
>
> I'm just a little concerned this work is getting a bit outside of the
> scope of the original idea of this PR, and I wonder if it would make sense
> to split it into smaller, separate PRs that each basically do one thing
> only. This will make the review process faster and easier, and allows for a
> better discussion of the individual changes you're proposing!
>
> For example, while I do like the idea of adding a shortcut for settings,
> it should conform to the system standards, e.g. Cmd+, on macOS, and to
> whatever the defaults might be in windows and Linux.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-qt-browser/pull/85#issuecomment-1091099719>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACEKMIG63SOLI7GMPT37RZTVDZX5TANCNFSM5SGB4NTQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>
-- 

--
Juan

agramfort commented on mne_qt_browser/tests/test_pg_specific.py:
```diff
@@ -4,6 +4,8 @@
 # License: BSD-3-Clause
 
 import numpy as np
+import math
```
```suggestion
import math
import numpy as np
```

import standard lib first

juangpc replied:
with later changes math lib no longer needed.

agramfort commented on mne_qt_browser/tests/test_pg_specific.py:
```diff
@@ -81,3 +83,281 @@ def test_annotations_interactions(raw_orig, pg_backend):
+        QTest.qWait(50)
+        sensitivity_test[i] = fig.mne.scroll_sensitivity == sensitivity_control.value()
+    assert all(sensitivity_test)
+
```
you should run flake8 on your code

juangpc replied:
done: 3aefc9d989feb151580cc639dc0b33acd70a3284

hoechenberger commented on mne_qt_browser/_pg_figure.py:
```diff
@@ -1697,10 +1697,10 @@ def __init__(self, main, **kwargs):
     """A dialog to toggle projections."""
 
-    def __init__(self, main, *, name):
+    def __init__(self, main, title='Projectors', **kwargs):
```
Why do we need the flexibility to adjust the title of that dialog?

Also, in #101 I specifically changed the signature to **not** accept `kwargs` and instead enforce passing keyword-only arguments to ensure better code health. Please revert the change in this line

agramfort replied:
this one also needs to be reverted. We really try to avoid kwargs in mne as it makes the code super hard to document and test.

jasmainak replied:
I spent some time looking at the code with @juangpc and I agree **kwargs made the code really hard to read. And it's nested really deep with the backend handling. What used to be "params" in the past is now "kwargs"

juangpc replied:
I see. Well. I'm sorry. I'll fix this.

> Why do we need the flexibility to adjust the title of that dialog?

I don't need it. I changed it back because: 1. coherence. 2. your code when improperly merged with mine, introduced a bug. 
Attending 1) there are 5 or so dialog windows child classes to _BaseDialog. And the way the child classes communicate with it in the other instances is different from what you proposed in your change. See SettingsDialog or HelpDialog, all go in the same toolbar.
Attending 2) The merge with your change introduced a bug in my fork. It might have been me or automatic merge. Don't remember, it was late. This has been solved now. I was surprised by your change, the tests found it. The tests I push here, actually found the bug. So I guess thats a good thing. Anyway it was a bad merge. I'll fix this.

Also, I get a warning when using ```name``` as argument in line 1700: "Shadows name 'name' from greater scope".

I don't know the difference between single star "*" and "**kwargs". This is my first python pr.

hoechenberger commented on mne_qt_browser/_pg_figure.py:
```diff
@@ -1697,10 +1697,10 @@ def __init__(self, main, **kwargs):
         self.external_change = True
         # Create projection-layout
-        super().__init__(main.window(), name=name, title='Projectors')
+        super().__init__(main.window(), title=title, **kwargs)
```
See above, please revert this

juangpc replied:
Attended 0408af1349194cae0016f598f2e85529846728b3

agramfort commented on mne_qt_browser/_pg_figure.py:
```diff
@@ -1697,10 +1697,10 @@ def __init__(self, main, **kwargs):
     """A dialog to toggle projections."""
 
-    def __init__(self, main, *, name):
+    def __init__(self, main, title='Projectors', **kwargs):
```
@juangpc you need to be careful when making large changes including to public functions. One PR should be on one feature / one bug.

Why do you need these changes to address the fix aimed at here? If it's not related to the fix can you revert the unrelated changes and put this in a follow up PR that can be discussed with the other devs?

juangpc replied:
> Why do you need these changes to address the fix aimed at here? 

I don't. Attended here: 0408af1349194cae0016f598f2e85529846728b3

agramfort commented on mne_qt_browser/_pg_figure.py:
```diff
@@ -1578,7 +1578,7 @@ def __init__(self, main, title='Settings', **kwargs):
                 'pyqtgraph)</i><br>'
                 'Default is "peak".')
-        self.ds_method_cmbx.addItems(['subsample', 'mean', 'peak'])
+        self.ds_method_cmbx.addItems(['mean', 'subsample', 'peak'])
```
this is unrelated right?

juangpc replied:
clearly unrelated to the crash. Attended here: 798c39d31e98c1ac1f64866f37c8ead470a2c994

juangpc commented:
Ready on my side.

larsoner commented on mne_qt_browser/tests/test_pg_specific.py:
```diff
@@ -81,3 +82,276 @@ def test_annotations_interactions(raw_orig, pg_backend):
+    if sensitivity_values[-1] != sensitivity_control.maximum():
+        sensitivity_values.append(sensitivity_control.maximum())
+    sensitivity_test = [False] * len(sensitivity_values)
+    for i in range(len(sensitivity_values)):
+        sensitivity_control.setValue(sensitivity_values[i])
+        QTest.qWait(50)
+        sensitivity_test[i] = \
+            fig.mne.scroll_sensitivity == sensitivity_control.value()
+    assert all(sensitivity_test)
```
Do you agree this is simpler? Pytest can handle list comparisons directly, so:
```suggestion
    sensitivity_mne = list()
    sensitivity_qt = list()
    for val in sensitivity_values:
        sensitivity_control.setValue(val)
        QTest.qWait(50)
        sensitivity_mne.append(fig.mne.scroll_sensitivity)
        sensitivity_qt.append(sensitivity_control.value())
    assert sensitivity_mne == sensitivity_values
    assert sensitivity_qt == sensitivity_values
```

larsoner commented on mne_qt_browser/tests/test_pg_specific.py:
```diff
@@ -81,3 +82,276 @@ def test_annotations_interactions(raw_orig, pg_backend):
+            fig.mne.scroll_sensitivity == sensitivity_control.value()
+    assert all(sensitivity_test)
+
+    sensitivity_test = [False] * len(sensitivity_values)
+    for i in reversed(range(len(sensitivity_values))):
+        sensitivity_control.setValue(sensitivity_values[i])
+        QTest.qWait(50)
+        sensitivity_test[i] = \
+            fig.mne.scroll_sensitivity == sensitivity_control.value()
+    assert all(sensitivity_test)
```
Same pattern here with
```
    ...
    for val in sensitivity_values[::-1]:
        ...
    assert sensitivity_mne == sensitivity_values[::-1]
    assert sensitivity_qt == sensitivity_values[::-1]
```

juangpc replied:
Yes. Minimal modification so that the code is almost identical to the previous section. See: e8d91bbbdfadbe1097974d3da821afb03a141859

larsoner commented on README.md:
````diff
@@ -98,10 +98,15 @@ Please report bugs and feature requests in the [issues](https://github.com/mne-t
 
 ```console
-pytest -m benchmark mne_qt_browser
+pytest -m benchmark mne_qt_browser/tests/test_speed.py
````
You shouldn't need to do this (I don't see much advantage)

juangpc replied:
I agree. Thanks. 
Changed here: 15c48136960bce8e102f5d50eeaa6786cbe37456

larsoner reviewed (commented):
Other than a couple of small suggestions, LGTM!

juangpc commented:
Ready on my side. 

Since this is getting close to ready for merging. Let me thank you all for your time. Hope this is useful.

agramfort commented:
@larsoner I let you merge if happy

larsoner commented:
Thanks @juangpc !
